Make it build with ghc-9.6

erikd opened the pull request:
This should not be merged until I can remove the `source-repository-package` stanzas that I added to the `cabal.project` file.

Also fixes #3439

erikd commented:
All the SRPs have been removed so this is not longer a draft/WIP.

lehins commented on cabal.project:
```diff
@@ -86,11 +87,13 @@ benchmarks: true
-  --sha256: 01fy4nbq6kaqi73ydn6w7rd1izkg5p217q5znyp2icybf41sl1b6
+constraints:
+    -- Earlier versions have a slightly different set of type constraints.
+    persistent >= 2.14
```
Can we put this either
* under the `if impl(ghc >= 9.6)` below?
* or directly in the `ledger-state` package?

`ledger-state` package is the only one using persist and it's not even released to CHaPs, so it is a low impact problem.

lehins commented on eras/byron/ledger/impl/src/Cardano/Chain/Update/Proposal.hs:
Could you please bump up the version of this ~`cardano-ledger-binary`~ `cardano-ledger-byron` package to `1.0.0.1` and a changelog entry saying GHC-9.6 compatibility.

This is a non-semantic change, thus a minor patch, but it is also a changelog worthy change.

lehins replied:
@erikd I am really sorry!!! I meant to say `cardano-ledger-byron`, not `cardano-ledger-binary`. My muscle memory messed with my brain.

erikd replied:
Done.

lehins commented on eras/shelley/test-suite/test/Test/Cardano/Ledger/Shelley/MultiSigExamples.hs:
```diff
@@ -179,11 +179,11 @@ makeTx ::
-        , scriptWits = msigs
-        }
+      mkBasicTxWits
+        & addrTxWitsL .~ (mkWitnessesVKey (hashAnnotated txBody) keyPairs)
```
Redundant parenthesis
```suggestion
        & addrTxWitsL .~ mkWitnessesVKey (hashAnnotated txBody) keyPairs
```

lehins reviewed (commented):
Just a couple of minor comments, otherwise looks great. @erikd thank you for doing this work.

erikd committed:
Make it build with ghc-9.6

erikd committed:
Delete libs/cardano-ledger-test/bench/Bench/Cardano/Ledger/Balance.hs

This is an old benchmark which does not even compile. Have to remove it from
the repo because `cabal-3.10` complains about its presence if its not
mentioned in the cabal file.

erikd committed:
Delete libs/non-integral/app/Main.hs

Not mentioned in the cabal file.

erikd committed:
CI: Add ghc-9.6.1 to the build matrix

erikd committed:
Bump cardano-ledger-binary version

angerman committed:
Nix: Bump hackage.nix and CHaP
